### PR TITLE
Adding CircularMapper as module to nf-core

### PR DESCRIPTION
**Pull request by shyama-mama**
## PR checklist

Opened as part of https://github.com/nf-core/eager/issues/1033
Adding CircularMapper to nf-core modules for lifter over of eager functionality from DSL1 to DSL2

- [x] This comment contains a description of changes (with reason).
- [x] If you've fixed a bug or added code that should be tested, add tests!
- [x] If you've added a new tool - have you followed the module conventions in the [contribution docs](https://github.com/nf-core/modules/tree/master/.github/CONTRIBUTING.md)
- [x] If necessary, include test data in your PR.
- [x] Remove all TODO statements.
- [x] Emit the `versions.yml` file.
- [x] Follow the naming conventions.
- [x] Follow the parameters requirements.
- [x] Follow the input/output options guidelines.
- [x] Add a resource `label`
- [x] Use BioConda and BioContainers if possible to fulfil software requirements.
- Ensure that the test works with either Docker / Singularity. Conda CI tests can be quite flaky:
  - [x] `PROFILE=docker pytest --tag <MODULE> --symlink --keep-workflow-wd --git-aware`
  - [ ] `PROFILE=singularity pytest --tag <MODULE> --symlink --keep-workflow-wd --git-aware`
  - [ ] `PROFILE=conda pytest --tag <MODULE> --symlink --keep-workflow-wd --git-aware`

### CODE REVIEW

**Comment by shyama-mama** [comment]
I've fixed the issues with prettier and formatting. But I'm not sure what is happening with pytests failing. The docker version I run locally finishes successfully.

File: modules/nf-core/circularmapper/circulargenerator/meta.yml [hunk 1]
```diff
@@ -0,0 +1,42 @@
...
+      documentation: "https://circularmapper.readthedocs.io/en/latest/index.html"
+      tool_dev_url: "https://github.com/apeltzer/CircularMapper/"
+      doi: "10.1186/s13059-016-0918-z"
+      licence: "['GPL v3']"
+input:
+  - meta:
+      type: map
+      description: |
+        Groovy Map containing reference information. Requires circularextension and circulartarget to function
+        e.g. `[ id:'test', circularextension:500, circulartarget:'chrMT' ]`
```
Comment (SPPearce) [inline modules/nf-core/circularmapper/circulargenerator/meta.yml:21]: Aha. These want to be passed as value channels. Do they have no defaults?

File: modules/nf-core/circularmapper/realignsamfile/main.nf [hunk 2]
```diff
@@ -0,0 +1,51 @@
+process CIRCULARMAPPER_REALIGNSAMFILE {
+    tag "$meta.id"
+    label 'process_medium'
+
+    conda "bioconda::circularmapper=1.93.5"
```
Comment (SPPearce) [inline modules/nf-core/circularmapper/realignsamfile/main.nf:5]: Same as before, now needs an environment file

File: modules/nf-core/circularmapper/realignsamfile/meta.yml [hunk 3]
```diff
@@ -0,0 +1,56 @@
...
+  - circular
+  - map
+  - reference
+  - fasta
+  - bam
+  - short-read
+  - bwa
+tools:
+  - "circularmapper":
+      description: "A method to improve mappings on circular genomes such as Mitochondira."
```
Comment (SPPearce) [inline modules/nf-core/circularmapper/realignsamfile/meta.yml:14]:
```suggestion
      description: "A method to improve mappings on circular genomes such as Mitochondria."
```

**Review by SPPearce — CHANGES_REQUESTED** [review]
Please take a look at [the guidelines on value channels](https://nf-co.re/docs/contributing/modules#general). In short, we don't want lots of extra required fields on meta maps for modules, we'd rather pass them as value channels.

File: modules/nf-core/circularmapper/circulargenerator/main.nf [hunk 4]
```diff
@@ -0,0 +1,53 @@
...
+    conda "bioconda::circularmapper=1.93.5"
+    container "${ workflow.containerEngine == 'singularity' && !task.ext.singularity_pull_docker_container ?
+        'https://depot.galaxyproject.org/singularity/circularmapper:1.93.5--h4a94de4_1':
+        'biocontainers/circularmapper:1.93.5--h4a94de4_1' }"
+
+    input:
+    tuple val(meta), path(fasta)
+
+    output:
+    tuple val(meta), path("*_${meta.circularextension}.${fasta.extension}") , emit: fasta
```
[the remark this answers is not in the document]
Comment (shyama-mama) [reply modules/nf-core/circularmapper/circulargenerator/main.nf]: I've added these to the args, and made it possible to update from the params. Let me know if this implementation makes sense.

**Comment by shyama-mama** [comment]
@SPPearce Sorry! I'm struggling to decrypt these errors again, again the pytest on docker works fine locally. Not sure why it is complaining about some bowtie2 file not being found.

**Comment by SPPearce** [comment]
> @SPPearce Sorry! I'm struggling to decrypt these errors again, again the pytest on docker works fine locally. Not sure why it is complaining about some bowtie2 file not being found.

Those are weird errors that I don't understand, but seem to be related to the pytest `tests/` files not being present in the git repo.
I'm rerunning the tests to see if that helps magically.
We are in the process of migrating from pytest to nf-test, with nf-test now being the standard, so might be something to do with that.

**Comment by shyama-mama** [comment]
If everything looks good to you @SPPearce  I'm happy for this to be merged.

File: modules/nf-core/circularmapper/circulargenerator/main.nf [hunk 5]
```diff
@@ -0,0 +1,51 @@
...
+    script:
+    def args = task.ext.args ?: ''
+    def prefix = task.ext.prefix ?: "${meta.id}"
+    def VERSION = '1.93.5' // WARN: Version information not provided by tool on CLI. Please update this string when bumping container versions.
+    """
+    circulargenerator \\
+        -Xmx${task.memory.toGiga()}g \\
+        $args \\
+        -i $fasta
+    mv ${fasta}_${prefix}_elongated ${fasta.baseName}_${prefix}.${fasta.extension}_${prefix}_elongated
```
Comment (SPPearce) [inline modules/nf-core/circularmapper/circulargenerator/main.nf:30]: Why does this have prefix twice?

File: tests/modules/nf-core/circularmapper/circulargenerator/nextflow.config [hunk 6]
```diff
@@ -0,0 +1,13 @@
+process {
+
+    publishDir = { "${params.outdir}/${task.process.tokenize(':')[-1].tokenize('_')[0].toLowerCase()}" }
+
+    withName: CIRCULARMAPPER_CIRCULARGENERATOR {
+        ext.args = { [
+            params.elongation_factor ? "-e ${params.elongation_factor}" : '-e 500',
```
Comment (SPPearce) [inline tests/modules/nf-core/circularmapper/circulargenerator/nextflow.config:7]: Are these required parameters to get the tool to run, with no default in the tool itself? 
We have a [policy of requiring mandatory inputs as an argument to the module](https://nf-co.re/docs/contributing/modules#migrating-from-pytest-to-nf-test).

**Review by SPPearce — COMMENTED** [review]
I think several of my comments still stand, if there are mandatory arguments to get the tool to run, they should be being given as value channels.

File: tests/modules/nf-core/circularmapper/circulargenerator/nextflow.config (hunk 6 above)
Comment (jbv2) [reply tests/modules/nf-core/circularmapper/circulargenerator/nextflow.config:7]: they don't have defaults and parameters are necesary

**Comment by SPPearce** [comment]
Hi @jbv2 , it looks like the module has been added in #5326, so I'm going to close this one.

**Pull request closed without merging**